### PR TITLE
global: addition of admin views

### PR DESCRIPTION
**Pull request by bruno314**
Signed-off-by: Bruno Cuc <bruno.cuc@cern.ch>

### CODE REVIEW

File: examples/app.py [hunk 1]
```diff
@@ -62,9 +63,28 @@
...
 FlaskOAuth(app)
 InvenioOAuthClient(app)
+
+
+def identity(x):
+    """."""
+    return x
+
+
+InvenioAdmin(app, permission_factory=identity, view_class_factory=identity)
```
Comment (jirikuncar) [inline examples/app.py]: `view_class_factory=lambda x: x`?

File: invenio_oauthclient/admin.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+from flask.ext.admin.contrib.sqla import ModelView
+from flask_babelex import gettext as _
+from wtforms.fields import TextAreaField
+
+from .models import RemoteAccount, RemoteToken, UserIdentity
+
+
+class RemoteAccountView(ModelView):
+    """Flask-Admin view to manage oauth."""
```
Comment (jirikuncar) [inline invenio_oauthclient/admin.py]: Please update docstring so it describes the view better.

File: invenio_oauthclient/admin.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+    column_labels = {
+        'id': 'ID',
+        'user_id': 'User ID',
+        'client_id': 'Client ID',
+    }
+
+
+class RemoteTokenView(ModelView):
+    """Flask-Admin view to manage oauth."""
```
Comment (jirikuncar) [inline invenio_oauthclient/admin.py]: Please update docstring so it describes the view better.

File: invenio_oauthclient/admin.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+    form_columns = (
+        'remote_account',
+        'token_type',
+        'access_token',
+        'secret',
+    )
+
+    column_labels = {
+        'id_remote_account': 'ID Remote Account',
```
Comment (jirikuncar) [inline invenio_oauthclient/admin.py]: It should be marked as translatable string with `_(...)`.

File: invenio_oauthclient/admin.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+from wtforms.fields import TextAreaField
+
+from .models import RemoteAccount, RemoteToken, UserIdentity
+
+
+class RemoteAccountView(ModelView):
+    """Flask-Admin view to manage oauth."""
+
+    can_view_details = True
+    can_delete = True
```
Comment (lnielsen) [inline invenio_oauthclient/admin.py]: Please disallow creation and editing of remote tokens/accounts

File: MANIFEST.in [hunk 6]
```diff
@@ -41,3 +41,4 @@ recursive-include docs Makefile
 recursive-include examples *.py
 recursive-include invenio_oauthclient *.html
 recursive-include invenio_oauthclient *.pot
+recursive-include invenio_oauthclient *.py
```
Comment (lnielsen) [inline MANIFEST.in]: Check if needed.

File: invenio_oauthclient/admin.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+    column_filters = ('id', 'user_id', 'client_id',)
+
+    column_default_sort = ('id', True)
+
+    column_display_all_relations = True
+    inline_models = (RemoteToken,)
+
+    column_labels = {
+        'id': 'ID',
```
Comment (lnielsen) [inline invenio_oauthclient/admin.py]: ``'ID'`` -> ``_('ID')``

File: invenio_oauthclient/models.py [hunk 8]
```diff
@@ -119,10 +118,18 @@ def delete(self):
...
+    def __repr__(self):
+        """Repr."""
+        return str(self.id)
+
 
 class RemoteToken(db.Model):
     """Storage for the access tokens for linked accounts."""
 
+    def __repr__(self):
+        """Repr."""
```
Comment (lnielsen) [inline invenio_oauthclient/models.py]: String representation fo model.

File: invenio_oauthclient/models.py [hunk 9]
```diff
@@ -119,10 +118,18 @@ def delete(self):
...
+        return str(self.token_type) + ' ' + str(self.access_token)
```
Comment (lnielsen) [inline invenio_oauthclient/models.py]: ``"{0} {1}".format(....)``

File: invenio_oauthclient/models.py [hunk 10]
```diff
@@ -142,7 +149,7 @@ class RemoteToken(db.Model):
     """Type of token."""
 
     access_token = db.Column(
-        TextEncryptedType(type_in=db.Text, key=_secret_key), nullable=False
+        TextEncryptedType(type_in=db.Text, key=_secret_key), nullable=True
```
Comment (lnielsen) [inline invenio_oauthclient/models.py]: Should not allow null.

File: invenio_oauthclient/models.py [hunk 11]
```diff
@@ -68,7 +67,7 @@ class RemoteAccount(db.Model):
     client_id = db.Column(db.String(255), nullable=False)
     """Client ID of remote application (defined in OAUTHCLIENT_REMOTE_APPS)."""
 
-    extra_data = db.Column(MutableDict.as_mutable(JSONType), nullable=False)
+    extra_data = db.Column(MutableDict.as_mutable(JSONType), nullable=True)
```
Comment (lnielsen) [inline invenio_oauthclient/models.py]: Ditto

File: invenio_oauthclient/admin.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+
+class RemoteTokenView(ModelView):
+    """Flask-Admin view to manage remote tokens from invenio-oauthclient."""
+
+    can_view_details = True
+
+    column_list = (
+        'id_remote_account',
+        'token_type',
+        'access_token',
```
Comment (lnielsen) [inline invenio_oauthclient/admin.py]: Access token has to be removed.

File: invenio_oauthclient/models.py (hunk 9 above)
Comment (jirikuncar) [inline invenio_oauthclient/models.py]: or better `"{0.token_type} {0.access_token}".format(self)`

File: invenio_oauthclient/admin.py (hunk 12 above)
Comment (jirikuncar) [inline invenio_oauthclient/admin.py]: ping @bruno314

File: invenio_oauthclient/ext.py [hunk 13]
```diff
@@ -29,7 +29,8 @@
 from flask_login import user_logged_out
 from flask_oauthlib.client import OAuth as FlaskOAuth
 
-from . import config
+from invenio_oauthclient import config
```
Comment (jirikuncar) [inline invenio_oauthclient/ext.py]: `from . import config`

File: invenio_oauthclient/models.py [hunk 14]
```diff
@@ -232,4 +239,5 @@ class UserIdentity(db.Model):
         db.Index('useridentity_id_user_method', id_user, method, unique=True),
     )
 
+
```
Comment (jirikuncar) [inline invenio_oauthclient/models.py]: not needed

File: invenio_oauthclient/models.py [hunk 15]
```diff
@@ -119,10 +118,18 @@ def delete(self):
...
+        """String representation for model."""
+        return str(self.id)
+
 
 class RemoteToken(db.Model):
     """Storage for the access tokens for linked accounts."""
 
+    def __repr__(self):
+        """String representation for model."""
+        return "{0.token_type} {0.access_token}".format(self)
```
Comment (jirikuncar) [inline invenio_oauthclient/models.py]: `Remote Token <type=..., access_token=...>`

File: setup.py [hunk 16]
```diff
@@ -160,6 +161,12 @@ def run_tests(self):
         'invenio_i18n.translations': [
             'invenio_oauthclient = invenio_oauthclient',
         ],
+        'invenio_admin.views': [
+          'invenio_oauth_remote_account = '
+          'invenio_oauthclient.admin:remote_account_adminview',
+          'invenio_oauth_remote_token = '
+          'invenio_oauthclient.admin:remote_token_adminview',
+        ]
```
Comment (jirikuncar) [inline setup.py]: missing `,` at the end of line

**Comment by jirikuncar** [comment]
- [ ] update copyright year in files you have touched

File: tests/test_admin.py [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+        request_url = url_for("remoteaccount.index_view")
+
+    with app.app_context():
+        with app.test_client() as client:
+            res = client.get(
+                request_url,
+                follow_redirects=True
+            )
+            assert res.status_code == 200
+            # import ipdb; ipdb.set_trace()
```
Comment (jirikuncar) [inline tests/test_admin.py]: !!!

**Pull request closed without merging**